src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -962,117 +1000,300 @@ internal void ExecuteSubmission(BuildSubmission submission, bool allowMainThread
             ErrorUtilities.VerifyThrowArgumentNull(submission, nameof(submission));
             ErrorUtilities.VerifyThrow(!submission.IsCompleted, "Submission already complete.");
 
-            lock (_syncLock)
+            if (ProjectCacheIsPresent())
```

Comment:
Start review here. This method is the entrypoint from where the logic starts diverting when a cache is available.